doc-content/enterprise-only/openshift/hacep-clientcode-ref.adoc

```diff
@@ -0,0 +1,43 @@
+[id='hacep-clientcode-ref']
+= Requirements for client code
+
+When developing client code for high-availability CEP, you must follow certain additional requirements.
+
+[float]
+== kie-remote API
+
+The code must use the kie-remote API and not the kie API. The kie-remote API is specified in the `org.kie:kie-remote` Maven artifact. You can find the source code in the `kie-remote` Maven module.
+
+[float]
+== Explicit timestamps
+
+The engine needs to determine the sequence in which events happen. For this reason every event must have an associated timestamp assigned to it. In a high-availability environment, make this timestamp a property of the Java bean modelling
+the event. You must then annotate the event class with `@Timestamp` annotation, where the name of the timestamp attribute itself is the parameter, as in the following example:
+
+[source,java]
+----
+@Role(Role.Type.EVENT)
+@Timestamp("myTime")
+public class StockTickEvent implements Serializable {
+
+    private String company;
+    private double price;
+    private long myTime;
+}
+----
+
+If you do not provide a timestamp attribute, Drools assigns a timestamp to every event based on the time  when the event is inserted by the client into a remote session. However, this mechanism depends on the clocks on the client machines. If clocks between different clients diverge, inconsistencies between events inserted by these hosts can happen.
+
+[float]
+== Lambda expressions for non-memory actions
+
+Memory actions (actions to insert, modify, or delete information in the working memory of the engine) also processed on every node of the cluster. Actions that are not memory actions must be executed only on the leader. For this reason, if any action that is not a memory action is required, it must be wrapped inside a lambda expression, for example:
```

Comment:
"working memory of the {DECISION_ENGINE}) ..."  and I assume should be "...) are also processed ..."

"is required, you must wrap the memory action inside..., as shown in the following example:"